File: android/src/main/java/com/margelo/nitro/nitroimage2pdf/PdfBuilder.kt
package com.margelo.nitro.nitroimage2pdf

import android.content.Context
import com.margelo.nitro.nitroimage2pdf.FitMode
import com.margelo.nitro.nitroimage2pdf.PageSizePreset
import com.margelo.nitro.nitroimage2pdf.PageConfig
import com.margelo.nitro.nitroimage2pdf.PdfMetadata
import com.tom_roush.pdfbox.android.PDFBoxResourceLoader
import com.tom_roush.pdfbox.pdmodel.PDDocument
import com.tom_roush.pdfbox.pdmodel.PDDocumentInformation
import com.tom_roush.pdfbox.pdmodel.PDPage
import com.tom_roush.pdfbox.pdmodel.PDPageContentStream
import com.tom_roush.pdfbox.pdmodel.common.PDRectangle
import com.tom_roush.pdfbox.pdmodel.encryption.AccessPermission
import com.tom_roush.pdfbox.pdmodel.encryption.StandardProtectionPolicy
import com.tom_roush.pdfbox.pdmodel.graphics.image.JPEGFactory
import com.tom_roush.pdfbox.pdmodel.graphics.image.LosslessFactory
import com.tom_roush.pdfbox.pdmodel.graphics.image.PDImageXObject
import java.io.ByteArrayOutputStream

// A single page's rendered size, in PDF points.
data class PdfPageSize(val width: Float, val height: Float)

// Bottom-left-origin box an image is drawn into.
private data class ContentBox(val x: Float, val y: Float, val width: Float, val height: Float)

// Assembles a multi-page PDF from LoadedImages via PdfBox-Android.
object PdfBuilder {
  // Standard page sizes in points (72 dpi), matching iOS.
  private const val A4_WIDTH = 595.28f
  private const val A4_HEIGHT = 841.89f
  private const val A5_WIDTH = 419.53f
  private const val A5_HEIGHT = 595.28f
  private const val LETTER_WIDTH = 612f
  private const val LETTER_HEIGHT = 792f
  private const val LEGAL_WIDTH = 612f
  private const val LEGAL_HEIGHT = 1008f

  // AES-128 is the max PdfBox-Android supports for writing.
  private const val ENCRYPTION_KEY_LENGTH = 128

  private const val DEFAULT_JPEG_QUALITY = 1.0f

  @Volatile
  private var initialized = false
  private val initLock = Any()

  // Must run once before any PDDocument work (idempotent).
  private fun ensureInit(context: Context) {
    if (initialized) return
    synchronized(initLock) {
      if (initialized) return
      PDFBoxResourceLoader.init(context.applicationContext)
      initialized = true
    }
  }

  // Resolves the page size in points for a given preset.
  private fun pageSize(
    preset: PageSizePreset,
    customWidth: Double?,
    customHeight: Double?,
    imageWidth: Int,
    imageHeight: Int
  ): PdfPageSize {
    return when (preset) {
      PageSizePreset.FIT -> PdfPageSize(imageWidth.toFloat(), imageHeight.toFloat())
      PageSizePreset.A4 -> PdfPageSize(A4_WIDTH, A4_HEIGHT)
      PageSizePreset.A5 -> PdfPageSize(A5_WIDTH, A5_HEIGHT)
      PageSizePreset.LETTER -> PdfPageSize(LETTER_WIDTH, LETTER_HEIGHT)
      PageSizePreset.LEGAL -> PdfPageSize(LEGAL_WIDTH, LEGAL_HEIGHT)
      PageSizePreset.CUSTOM -> PdfPageSize(
        (customWidth ?: imageWidth.toDouble()).toFloat(),
        (customHeight ?: imageHeight.toDouble()).toFloat()
      )
    }
  }

  // Content box honoring margin + fit; caller clips for cover.
  private fun contentBox(pageSize: PdfPageSize, imageWidth: Float, imageHeight: Float, fitMode: FitMode, margin: Float): ContentBox {
    val boxX = margin
    val boxY = margin
    val boxWidth = (pageSize.width - 2 * margin).coerceAtLeast(0f)
    val boxHeight = (pageSize.height - 2 * margin).coerceAtLeast(0f)

    if (imageWidth <= 0f || imageHeight <= 0f) {
      return ContentBox(boxX, boxY, boxWidth, boxHeight)
    }

    return when (fitMode) {
      FitMode.STRETCH -> ContentBox(boxX, boxY, boxWidth, boxHeight)
      FitMode.CONTAIN -> {
        val scale = minOf(boxWidth / imageWidth, boxHeight / imageHeight)
        val width = imageWidth * scale
        val height = imageHeight * scale
        ContentBox(boxX + (boxWidth - width) / 2, boxY + (boxHeight - height) / 2, width, height)
      }
      FitMode.COVER -> {
        val scale = maxOf(boxWidth / imageWidth, boxHeight / imageHeight)
        val width = imageWidth * scale
        val height = imageHeight * scale
        ContentBox(boxX + (boxWidth - width) / 2, boxY + (boxHeight - height) / 2, width, height)
      }
    }
  }

  // AccessPermission() defaults grant everything; deny print/modify/copy explicitly.
  private fun restrictedAccessPermission(): AccessPermission {
    return AccessPermission().apply {
      setCanPrint(false)
      setCanPrintDegraded(false)
      setCanPrintFaithful(false)
      setCanModify(false)
      setCanModifyAnnotations(false)
      setCanFillInForm(false)
      setCanAssembleDocument(false)
      setCanExtractContent(false)
    }
  }

  // Parses #RRGGBB into (r, g, b) floats 0..1, defaulting to white.
  private fun backgroundColor(hex: String?): Triple<Float, Float, Float> {
    val value = hex?.removePrefix("#")
    if (value == null || value.length != 6) return Triple(1f, 1f, 1f)
    val parsed = value.toIntOrNull(16) ?: return Triple(1f, 1f, 1f)
    val r = ((parsed shr 16) and 0xFF) / 255f
    val g = ((parsed shr 8) and 0xFF) / 255f
    val b = (parsed and 0xFF) / 255f
    return Triple(r, g, b)
  }

  // Builds the PDF bytes plus each page's size in points.
  fun build(
    context: Context,
    images: List<LoadedImage>,
    page: PageConfig?,
    metadata: PdfMetadata?,
    quality: Double?,
    userPassword: String?,
    ownerPassword: String?
  ): Pair<ByteArray, List<PdfPageSize>> {
    if (images.isEmpty()) {
      throw PdfException.InvalidImage("No images were provided to render")
    }

    ensureInit(context)

    val preset = page?.size ?: PageSizePreset.FIT
    val fitMode = page?.fitMode ?: FitMode.CONTAIN
    val margin = (page?.margin ?: 0.0).toFloat()
    val customWidth = page?.customWidth
    val customHeight = page?.customHeight
    val background = backgroundColor(page?.backgroundColor)
    val jpegQuality = (quality ?: 1.0).toFloat().coerceIn(0f, 1f)

    val document = PDDocument()
    try {
      val pageSizes = mutableListOf<PdfPageSize>()

      for (image in images) {
        val bitmap = image.bitmap
        val size = pageSize(preset, customWidth, customHeight, bitmap.width, bitmap.height)
        pageSizes.add(size)

        val pdfPage = PDPage(PDRectangle(size.width, size.height))
        document.addPage(pdfPage)

        val pdImage: PDImageXObject = try {
          if (image.isJpeg) {
            JPEGFactory.createFromImage(document, bitmap, jpegQuality)
          } else {
            LosslessFactory.createFromImage(document, bitmap)
          }
        } catch (e: Exception) {
          throw PdfException.InvalidImage("Failed to encode image for PDF: ${e.message}")
        }

        try {
          PDPageContentStream(document, pdfPage).use { contentStream ->
            // Background must be filled before the image, always.
            contentStream.setNonStrokingColor(background.first, background.second, background.third)
            contentStream.addRect(0f, 0f, size.width, size.height)
            contentStream.fill()

            val box = contentBox(size, bitmap.width.toFloat(), bitmap.height.toFloat(), fitMode, margin)

            when (fitMode) {
              FitMode.COVER -> {
                // cover overflows by design — clip to avoid margin bleed.
                contentStream.saveGraphicsState()
                val clipWidth = (size.width - 2 * margin).coerceAtLeast(0f)
                val clipHeight = (size.height - 2 * margin).coerceAtLeast(0f)
                contentStream.addRect(margin, margin, clipWidth, clipHeight)
                contentStream.clip()
                contentStream.drawImage(pdImage, box.x, box.y, box.width, box.height)
                contentStream.restoreGraphicsState()
              }
              FitMode.CONTAIN, FitMode.STRETCH -> {
                contentStream.drawImage(pdImage, box.x, box.y, box.width, box.height)
              }
            }
          }
        } catch (e: PdfException) {
          throw e
        } catch (e: Exception) {
          throw PdfException.WriteFailed("Failed to render page ${pageSizes.size}: ${e.message}")
        }
      }

      val info = PDDocumentInformation()
      metadata?.title?.let { info.title = it }
      metadata?.author?.let { info.author = it }
      metadata?.subject?.let { info.subject = it }
      metadata?.keywords?.let { info.keywords = it }
      metadata?.creator?.let { info.creator = it }
      document.documentInformation = info

      if (!userPassword.isNullOrEmpty() || !ownerPassword.isNullOrEmpty()) {
        try {
          // Restrict permissions only when there's no user/open password.
          val ownerPasswordOnly = !ownerPassword.isNullOrEmpty() && userPassword.isNullOrEmpty()
          val permissions = if (ownerPasswordOnly) restrictedAccessPermission() else AccessPermission()
          val policy = StandardProtectionPolicy(ownerPassword ?: userPassword ?: "", userPassword ?: "", permissions)
          policy.encryptionKeyLength = ENCRYPTION_KEY_LENGTH
          policy.isPreferAES = true
          document.protect(policy)
        } catch (e: Exception) {
          throw PdfException.EncryptionFailed("Failed to encrypt PDF: ${e.message}")
        }
      }

      val output = ByteArrayOutputStream()
      try {
        document.save(output)
      } catch (e: Exception) {
        throw PdfException.WriteFailed("Failed to serialize PDF: ${e.message}")
      }

      return output.toByteArray() to pageSizes
    } finally {
      document.close()
    }
  }
}

File: ios/PdfBuilder.swift
// Assembles a multi-page PDF from UIImages via UIGraphicsPDFRenderer.

import UIKit

enum PdfBuildError: Error {
  case emptyImages
}

// CustomStringConvertible keeps the [CODE] prefix consistent with PdfError.
extension PdfBuildError: CustomStringConvertible {
  var description: String {
    switch self {
    case .emptyImages:
      return "[INVALID_IMAGE] No images were provided to render"
    }
  }
}

extension PdfBuildError: LocalizedError {
  var errorDescription: String? { description }
}

enum PdfBuilder {
  // Standard page sizes in points (72 dpi).
  private static let a4Size = CGSize(width: 595.28, height: 841.89)
  private static let a5Size = CGSize(width: 419.53, height: 595.28)
  private static let letterSize = CGSize(width: 612, height: 792)
  private static let legalSize = CGSize(width: 612, height: 1008)

  // Resolves the page size in points for a given preset.
  static func pageSize(
    for preset: PageSizePreset,
    customWidth: Double?,
    customHeight: Double?,
    image: UIImage
  ) -> CGSize {
    switch preset {
    case .fit:
      return image.size
    case .a4:
      return a4Size
    case .a5:
      return a5Size
    case .letter:
      return letterSize
    case .legal:
      return legalSize
    case .custom:
      return CGSize(
        width: customWidth ?? image.size.width,
        height: customHeight ?? image.size.height
      )
    }
  }

  // Content rect honoring margin + fit; caller clips for cover.
  static func contentRect(pageSize: CGSize, imageSize: CGSize, fitMode: FitMode, margin: CGFloat) -> CGRect {
    let box = CGRect(
      x: margin,
      y: margin,
      width: max(pageSize.width - 2 * margin, 0),
      height: max(pageSize.height - 2 * margin, 0)
    )

    guard imageSize.width > 0, imageSize.height > 0 else { return box }

    switch fitMode {
    case .stretch:
      return box
    case .contain:
      let scale = min(box.width / imageSize.width, box.height / imageSize.height)
      let width = imageSize.width * scale
      let height = imageSize.height * scale
      return CGRect(x: box.midX - width / 2, y: box.midY - height / 2, width: width, height: height)
    case .cover:
      let scale = max(box.width / imageSize.width, box.height / imageSize.height)
      let width = imageSize.width * scale
      let height = imageSize.height * scale
      return CGRect(x: box.midX - width / 2, y: box.midY - height / 2, width: width, height: height)
    }
  }

  // Parses a #RRGGBB hex string, defaulting to white.
  static func backgroundColor(fromHex hex: String?) -> UIColor {
    guard var hexString = hex else { return .white }
    if hexString.hasPrefix("#") { hexString.removeFirst() }
    guard hexString.count == 6, let value = UInt32(hexString, radix: 16) else { return .white }

    let red = CGFloat((value >> 16) & 0xFF) / 255
    let green = CGFloat((value >> 8) & 0xFF) / 255
    let blue = CGFloat(value & 0xFF) / 255
    return UIColor(red: red, green: green, blue: blue, alpha: 1)
  }

  // jpegDataProviderSource preserves JPEG compression when embedded in the PDF.
  private static func applyQuality(_ image: UIImage, quality: Double) -> UIImage {
    let flattened: UIImage
    if image.imageOrientation == .up {
      flattened = image
    } else {
      let renderer = UIGraphicsImageRenderer(size: image.size)
      flattened = renderer.image { _ in image.draw(at: .zero) }
    }
    guard let jpegData = flattened.jpegData(compressionQuality: CGFloat(quality)),
          let dataProvider = CGDataProvider(data: jpegData as CFData),
          let cgImage = CGImage(
            jpegDataProviderSource: dataProvider,
            decode: nil,
            shouldInterpolate: true,
            intent: .defaultIntent
          )
    else {
      return image // fall back to the original if re-encoding somehow fails
    }
    return UIImage(cgImage: cgImage, scale: 1, orientation: .up)
  }

  // Builds the PDF Data plus each page's size in points.
  static func build(
    images: [UIImage],
    page: PageConfig?,
    processing: ProcessingConfig?,
    metadata: PdfMetadata?,
    userPassword: String?,
    ownerPassword: String?
  ) throws -> (data: Data, pageSizes: [CGSize]) {
    guard !images.isEmpty else { throw PdfBuildError.emptyImages }

    // quality == 1 (default) skips compression entirely.
    let quality = processing?.quality.map { min(max($0, 0), 1) }
    let images = quality.map { q -> [UIImage] in
      guard q < 1 else { return images }
      return images.map { applyQuality($0, quality: q) }
    } ?? images

    let preset = page?.size ?? .fit
    let fitMode = page?.fitMode ?? .contain
    let margin = CGFloat(page?.margin ?? 0)
    let customWidth = page?.customWidth
    let customHeight = page?.customHeight
    let background = backgroundColor(fromHex: page?.backgroundColor)

    var documentInfo: [String: Any] = [:]
    if let title = metadata?.title { documentInfo[kCGPDFContextTitle as String] = title }
    if let author = metadata?.author { documentInfo[kCGPDFContextAuthor as String] = author }
    if let subject = metadata?.subject { documentInfo[kCGPDFContextSubject as String] = subject }
    if let keywords = metadata?.keywords { documentInfo[kCGPDFContextKeywords as String] = keywords }
    if let creator = metadata?.creator { documentInfo[kCGPDFContextCreator as String] = creator }
    if let userPassword { documentInfo[kCGPDFContextUserPassword as String] = userPassword }
    if let ownerPassword { documentInfo[kCGPDFContextOwnerPassword as String] = ownerPassword }
    if userPassword != nil || ownerPassword != nil {
      // RC4 128-bit is Core Graphics' max encryption strength.
      documentInfo[kCGPDFContextEncryptionKeyLength as String] = 128
    }

    let format = UIGraphicsPDFRendererFormat()
    format.documentInfo = documentInfo

    let firstPageSize = pageSize(for: preset, customWidth: customWidth, customHeight: customHeight, image: images[0])
    let renderer = UIGraphicsPDFRenderer(bounds: CGRect(origin: .zero, size: firstPageSize), format: format)

    var pageSizes: [CGSize] = []
    let data = renderer.pdfData { context in
      for image in images {
        let size = pageSize(for: preset, customWidth: customWidth, customHeight: customHeight, image: image)
        pageSizes.append(size)

        context.beginPage(withBounds: CGRect(origin: .zero, size: size), pageInfo: [:])
        let cgContext = context.cgContext

        // Background must be filled before the image, always.
        background.setFill()
        cgContext.fill(CGRect(origin: .zero, size: size))

        let rect = contentRect(pageSize: size, imageSize: image.size, fitMode: fitMode, margin: margin)

        switch fitMode {
        case .cover:
          // cover overflows by design — clip to avoid margin bleed.
          let clipRect = CGRect(
            x: margin,
            y: margin,
            width: max(size.width - 2 * margin, 0),
            height: max(size.height - 2 * margin, 0)
          )
          cgContext.saveGState()
          cgContext.clip(to: clipRect)
          image.draw(in: rect) // UIImage.draw(in:) auto-applies EXIF orientation.
          cgContext.restoreGState()
        case .contain, .stretch:
          image.draw(in: rect)
        }
      }
    }

    return (data, pageSizes)
  }
}
